src/commonMain/kotlin/exchange.dydx.abacus/calculator/TradeInputCalculator.kt

```diff
@@ -1545,30 +1537,20 @@ internal class TradeInputCalculator(
             "STOP_MARKET", "TAKE_PROFIT_MARKET" -> {
                 parser.asNativeMap(trade["marketOrder"])?.let { marketOrder ->
                     val feeRate = parser.asDouble(parser.value(user, "takerFeeRate"))
-                    val bestPrice = marketOrderBestPrice(marketOrder)
+                    val midMarketPrice = marketOrderbookMidPrice(market)
                     val worstPrice = marketOrderWorstPrice(marketOrder)
-                    val slippageStepSize = 0.00001
-                    val slippage =
-                        if (worstPrice != null && bestPrice != null && bestPrice > Numeric.double.ZERO) {
-                            Rounder.quickRound(
-                                (worstPrice - bestPrice).abs() / bestPrice,
-                                slippageStepSize,
-                            )
-                        } else {
-                            null
-                        }
+                    val slippageFromMidPrice = marketOrderSlippageFromMidPrice(worstPrice, midMarketPrice)
 
-                    val triggerPrice =
-                        parser.asDouble(parser.value(trade, "price.triggerPrice"))
+                    val triggerPrice = parser.asDouble(parser.value(trade, "price.triggerPrice"))
                     val marketOrderPrice = marketOrderPrice(marketOrder)
-                    val priceslippage =
-                        if (bestPrice != null && marketOrderPrice != null) (marketOrderPrice - bestPrice) else null
+                    val priceSlippage = if (midMarketPrice != null && marketOrderPrice != null) (marketOrderPrice - midMarketPrice) else null
                     val slippagePercentage =
-                        if (priceslippage != null && bestPrice != null) {
-                            abs(priceslippage / bestPrice)
+                        if (priceSlippage != null && midMarketPrice != null) {
```

Comment:
nit, can we combine the above two lines with this to jsut be 

```
val slippagePercentage = if (priceSlippage != null && midMarketPrice != null) {
  abs( (marketOrderPrice - midMarketPrice) / midMarketPrice ) 
} else { null } 
```

just to avoid redundant null checks

Reply:
also, should we safeguard against `midMarketPrice = 0`? I imagine it's pretty unlikely but still possible

Reply:
umm unlikely since ask/bid price can never be 0 or negative, but i guess if the rounding goes very wrong x_x? will add!